### ISSUE TITLE
Migrate to new lifecycle methods

### ISSUE THREAD

**Issue by yesmeck** [ant-design/ant-design#9792]
React 16.3 will introduce a StrictMode which will fires deprecation warnings for `componentWillMount`, `componentWillUpdate`, `componentWillReceiveProps`. We should migrate antd's components to the new API, ~[react-af](https://github.com/donavon/react-af) can be a good start.~

- Example use antd with StrictMode https://codesandbox.io/s/xvrw82m6qw
- React 16.3 https://github.com/facebook/react/issues/12152

Here is a todo list, if you have interest in helping, drop a comment below.

- [x] Alert
- [x] Anchor @natanielcz
- [x] AutoComplete maybe this one needn't to update?
- [x] Avatar
- [x] BackTop
- [x] Badge
- [x] Breadcrunmb
- [x] Button @natanielcz
- [x] Calendar @natanielcz
- [x] Card
- [x] Carousel https://github.com/akiran/react-slick/commit/97988e897750e1d8f7b10a86b655f50d75d38298
- [x] Cascader
- [x] Checkbox
- [x] Collapse https://github.com/react-component/collapse/pull/129
- [x] DatePicker
- [x] Divider
- [x] Dropdown @klymenko-serhii
- [x] Form waitting for upgrade of [form](https://github.com/react-component/form) https://github.com/react-component/animate/pull/82
- [x] Grid
- [x] Icon
- [x] Input @natanielcz
- [x] InputNumber https://github.com/react-component/input-number/pull/194
- [x] Layout
- [x] List
- [x] LocaleProvider
- [x] Mention
- [x] Menu @picodoth https://github.com/react-component/menu/pull/147
- [x] Message waitting for upgrade of [animate](https://github.com/react-component/animate)
- [x] Modal https://github.com/react-component/dialog/pull/135
- [x] Notification waitting for upgrade of [animate](https://github.com/react-component/animate)
- [x] Pagination waitting for upgrade of [pagination](https://github.com/react-component/pagination)
- [x] Popconfirm
- [x] Popover
- [x] Progress maybe this one needn't to update?
- [x] Radio waitting for upgrade of [checkbox](https://github.com/react-component/checkbox)
- [x] Rate https://github.com/react-component/rate/pull/18
- [x] Select @yutingzhao1991 https://github.com/react-component/select/pull/291
- [x] Slider waitting for https://github.com/react-component/slider/pull/591
- [x] Spin waitting for upgrade of [animate](https://github.com/react-component/animate)
- [x] Steps
- [x] Switch https://github.com/react-component/switch/pull/36
- [x] Table @yesmeck 
   - [x] https://github.com/react-component/table/pull/198
   - [x] https://github.com/ant-design/ant-design/pull/18266
   - [x] https://github.com/ant-design/ant-design/pull/18250
- [x] Tabs https://github.com/react-component/tabs/pull/213
- [x] Tag
- [x] TimePicker https://github.com/react-component/time-picker/pull/164
- [x] Timeline maybe this one needn't to update?
- [x] Tooltip
- [x] Transfer waitting for upgrade of Input, Button, [animate](https://github.com/react-component/animate)
- [x] Tree @zombieJ
- [x] TreeSelect @zombieJ
- [x] Upload

**Comment by afc163** [comment]
16.3

**Comment by yesmeck** [comment]
https://github.com/reactjs/react-lifecycles-compat

**Comment by yesmeck** [comment]
[Update on Async Rendering](https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html)

**Comment by chenshuai2144** [comment]
正在在搞 select  

我要 select

**Comment by saostad** [comment]
Guys I am using antd in one of my projects and it's awesome, How can I do anything regarding to this issue
?

**Comment by yesmeck** [comment]
@saostad You can pick up a component, and rewrite the legacy lifecycle methods with the new lifecycle method, see this PR https://github.com/react-component/table/pull/198 as a example.

**Comment by zombieJ** [comment]
add me on `Tree` thanks.

**Comment by yesmeck** [comment]
https://github.com/reactjs/rfcs/pull/40

**Comment by Rohanhacker** [comment]
Can I work on datepicker and modal ?? @yesmeck

**Comment by picodoth** [comment]
I'll be working on `menu`

**Comment by yesmeck** [comment]
@Rohanhacker Of course.

**Comment by jzhangs** [comment]
on `Tabs`

**Comment by klymenko-serhii** [comment]
ill try dropdown

**Comment by paranoidjk** [comment]
More version detect and compatibility code....

**Comment by picodoth** [comment]
taking tabs

**Comment by viztor** [comment]
interested in the <message>.

**Comment by ahsannaveed007** [comment]
Hi, is this still in progress❓

**Comment by dayrlism10** [comment]
ahh, is there anyway to prevent the legacy error pops in the console before this all finishing, seems like the progress is abit slow :(
i cant update to the latest and use new features like drawer as it pop me a lots of error.

**Comment by yesmeck** [comment]
@dayrlism10 What's the error?

**Comment by dayrlism10** [comment]
![legacy_code_error](https://user-images.githubusercontent.com/38413388/43573393-353b2a7a-9674-11e8-9647-eebcb025058b.png)

**Comment by yesmeck** [comment]
@dayrlism10 Are you using `<StrictMode>`?

**Comment by dayrlism10** [comment]
yes im using strict mode.

**Comment by yesmeck** [comment]
You can help us move to the new lifecycle methods if you want to use strict mode.

**Comment by dayrlism10** [comment]
can i know how to do so please?

**Comment by yesmeck** [comment]
Contributing guide https://ant.design/docs/react/contributing
New lifecycle methods https://reactjs.org/blog/2018/03/29/react-v-16-3.html
Example Pull Request: https://github.com/ant-design/ant-design/pull/10309

**Comment by dancerphil** [comment]
Trying migrate `Calendar`.
The `mode` of `Calendar` is neither `fully-controlled` nor `uncontrolled`.
The migration may cause behavior changes.
I suppose that `mode` is fully-controlled if `mode` is passed as props. And it is uncontrolled if it is not passed.

**Comment by Kexin-Li** [comment]
Hi, Can I work on Rate?

**Comment by natanielcz** [comment]
I think that **alert** has nothing to migrate

**Comment by IssuehuntBot** [comment]
@natanielcz has started working. [Visit this issue on Issuehunt](https://issuehunt.io/repos/34526884/issues/9792)

**Comment by natanielcz** [comment]
I try to rewrite all of them. Are our tests covering component lifecycle?

**Comment by yesmeck** [comment]
@natanielcz You can see coverage here https://codecov.io/gh/ant-design/ant-design

**Comment by yesmeck** [comment]
@Kexin-Li It's yours.

**Comment by dancerphil** [comment]
@yesmeck @natanielcz The tests right now are not enough to cover all the migration, you'd better add necessary tests before rewrite.

**Comment by yesmeck** [comment]
The list above includes rc-components, we'd have to rewrite both rc-components and antd componentns.

**Comment by yesmeck** [comment]
Updated the list.

**Comment by natanielcz** [comment]
I will create PR for each component so You could decide when to merge specific change. 
And It also will be easier to review smaller chunks of code.

**Comment by natanielcz** [comment]
Autocomplete, Carousel, Collapse, Form, InputNumber, Message, Modal, Notification, Pagination, Progress, Slider, Spin, Switch, Tabs and Timeline don't require migration.

**Comment by damiangreen** [comment]
Is there anything left that needs migrating?

**Comment by natanielcz** [comment]
Yeap. There are still components to migrate. I'm waiting for approval of first 5 and then I plan to continue work on this bug. If you want to help start from the end of the list above :)

**Comment by afc163** [comment]
Rate is done.

**Comment by Rohanhacker** [comment]
maybe we should wait for version 17 or hooks ??

**Comment by saostad** [comment]
> maybe we should wait for version 17 or hooks ??

these changes required for almost all new features, specially for React lazy, Suspense and Concurrent React.

**Comment by damiangreen** [comment]
Migrating ```transfer``` is tricky because ```componentWillReceiveProps```  refers to an instance property ```splitedDataSource``` which is used for caching..

**Comment by damiangreen** [comment]
Migrating ```mention```, the test ```'change suggestions'``` [here](https://github.com/ant-design/ant-design/blob/5dc5ab13bc788fc9bd67c4939db7bafe59cfb05e/components/mention/__tests__/index.test.js#L56) has a problem in that ```wrapper.setState``` doesn't set the state correctly when a new  ```getDerivedStateFromProps``` method is introduced. 

Looking closer at this test I'm wondering how useful is, because ```wrapper.setState``` is simulating an internal change, not an operation the user could ever do. The ```Suggestions``` array  should only be changed by modifying props. @afc163 do you have any thoughts about changing this test to remove wrapper.setstate()?

**Comment by IssuehuntBot** [comment]
@0maxxam0 funded this issue with $20. [See it on IssueHunt](https://issuehunt.io/repos/34526884/issues/9792)

**Comment by damiangreen** [comment]
If the PRs above are merged there is one component left to migrate (```transfer```). Out of curiosity does anyone know how you submit multiple PRs to issuehunt?

**Comment by zombieJ** [comment]
Cascader is done

**Comment by damiangreen** [comment]
@zombieJ  they are all done, except transfer, can you merge the PRs above?

**Comment by zy410419243** [comment]
I'll try `Button`.

**Comment by damiangreen** [comment]
@zy410419243  @natanielcz  has already done this one. We're just waiting on the maintainers to accept the PRs

**Comment by zombieJ** [comment]
@damiangreen, thanks for tips. Let me check.

**Comment by damiangreen** [comment]
@zombieJ, @afc163  what woudl it take to become a member/ mantainer. It seems you guys are super busy with so many pull requests and a few of us would be happy to help out?

**Comment by damiangreen** [comment]
brilliant, thanks @afc163

**Comment by IssuehuntBot** [comment]
@boostio has cancelled funding for this issue.(Cancelled amount: $30.00) [See it on IssueHunt](https://issuehunt.io/repos/34526884/issues/9792)

**Comment by IssuehuntBot** [comment]
@0maxxam0 has cancelled funding for this issue.(Cancelled amount: $20.00) [See it on IssueHunt](https://issuehunt.io/repos/34526884/issues/9792)

**Comment by heavenshell** [comment]
Any progress? 😅 

I Updated React 16.9.0, I got following warnings.

```
react-dom.development.js:11494 Warning: componentWillReceiveProps has been renamed, and is not recommended for use. See https://fb.me/react-async-component-lifecycle-hooks for details.

* Move data fetching code or side effects to componentDidUpdate.
* If you're updating state whenever props change, refactor your code to use memoization techniques or move it to static getDerivedStateFromProps. Learn more at: https://fb.me/react-derived-state
* Rename componentWillReceiveProps to UNSAFE_componentWillReceiveProps to suppress this warning in non-strict mode. In React 17.x, only the UNSAFE_ name will work. To rename all deprecated lifecycles to their new names, you can run `npx react-codemod rename-unsafe-lifecycles` in your project source folder.

Please update the following components: Tabs
```

**Comment by lPaths** [comment]
> Any progress? 
> 
> After I Updated React 16.9.0, I've got following warnings.
> 
> ```
> react-dom.development.js:11494 Warning: componentWillReceiveProps has been renamed, and is not recommended for use. See https://fb.me/react-async-component-lifecycle-hooks for details.
> 
> * Move data fetching code or side effects to componentDidUpdate.
> * If you're updating state whenever props change, refactor your code to use memoization techniques or move it to static getDerivedStateFromProps. Learn more at: https://fb.me/react-derived-state
> * Rename componentWillReceiveProps to UNSAFE_componentWillReceiveProps to suppress this warning in non-strict mode. In React 17.x, only the UNSAFE_ name will work. To rename all deprecated lifecycles to their new names, you can run `npx react-codemod rename-unsafe-lifecycles` in your project source folder.
> 
> Please update the following components: Tabs
> ```

AntForm is having same warnings

**Comment by Haqverdi** [comment]
Table also have this warnings after uprade react to 16.9

**Comment by baitun** [comment]
Hey, @afc163!
Can't we just run `npx react-codemod rename-unsafe-lifecycles` in antd project, or manually add prefix `UNSAFE_` for all unsafe methods, as said [here](https://reactjs.org/blog/2019/08/08/react-v16.9.0.html) to get rid of warnings?

**Comment by afc163** [comment]
PR is welcome guys~

**Comment by afc163** [comment]
react-slick: https://github.com/ant-design/ant-design/pull/18209/commits/b0cc0f7019f069d7a7c6a296ccc7ada3591c647d

**Comment by hLinx** [comment]
> > Any progress?
> > After I Updated React 16.9.0, I've got following warnings.
> > ```
> > react-dom.development.js:11494 Warning: componentWillReceiveProps has been renamed, and is not recommended for use. See https://fb.me/react-async-component-lifecycle-hooks for details.
> > 
> > * Move data fetching code or side effects to componentDidUpdate.
> > * If you're updating state whenever props change, refactor your code to use memoization techniques or move it to static getDerivedStateFromProps. Learn more at: https://fb.me/react-derived-state
> > * Rename componentWillReceiveProps to UNSAFE_componentWillReceiveProps to suppress this warning in non-strict mode. In React 17.x, only the UNSAFE_ name will work. To rename all deprecated lifecycles to their new names, you can run `npx react-codemod rename-unsafe-lifecycles` in your project source folder.
> > 
> > Please update the following components: Tabs
> > ```
> 
> AntForm is having same warnings


this problem will only happen to react v16.9
**through source code**

[ react v16.8.6](https://github.com/facebook/react/blob/16.8.6/packages/shared/ReactFeatureFlags.js)
`export const warnAboutDeprecatedLifecycles = false;`

[ react v16.9.0](https://github.com/facebook/react/blob/master/packages/shared/ReactFeatureFlags.js)
`export const warnAboutDeprecatedLifecycles = true;`

So the best solution now is to use react v16.8.6

**Comment by afc163** [comment]
https://weibo.com/1651843872/I2kqHzjDJ

**Comment by ZWkang** [comment]
[react-component/collapse#129](https://github.com/react-component/collapse/pull/129)

migrate lifecycle method with collapse component. any review ?

**Comment by XianZhengquan** [comment]
为什么不使用 react 推荐的 `npx react-codemod rename-unsafe-lifecycles` 修改呢？

**Comment by chenshuai2144** [comment]
XianZhengquan 治标不治本

**Comment by XianZhengquan** [comment]
@chenshuai2144 所以，现在你们是在慢慢的修改声明周期吗？ componentWillMount => componentDidMount ?

**Comment by chenshuai2144** [comment]
@XianZhengquan 你可以看一下写完的，其实都是重写组件，工作量还挺大的

**Comment by XianZhengquan** [comment]
@chenshuai2144 辛苦！4.0 预计什么时候完工呢？

**Comment by Stephen2** [comment]
ouch.  This looks messy.  Are we blocked from using React 16.9 or can we ignore warning?

**Comment by zombieJ** [comment]
@Stephen2 , we are working on the 4.0 which will all use new life cycle method. For current 3.x, may more wish social volunteer to help on this.

**Comment by xiaochouyu2663** [comment]
> > > Any progress?
> > > After I Updated React 16.9.0, I've got following warnings.
> > > ```
> > > react-dom.development.js:11494 Warning: componentWillReceiveProps has been renamed, and is not recommended for use. See https://fb.me/react-async-component-lifecycle-hooks for details.
> > > 
> > > * Move data fetching code or side effects to componentDidUpdate.
> > > * If you're updating state whenever props change, refactor your code to use memoization techniques or move it to static getDerivedStateFromProps. Learn more at: https://fb.me/react-derived-state
> > > * Rename componentWillReceiveProps to UNSAFE_componentWillReceiveProps to suppress this warning in non-strict mode. In React 17.x, only the UNSAFE_ name will work. To rename all deprecated lifecycles to their new names, you can run `npx react-codemod rename-unsafe-lifecycles` in your project source folder.
> > > 
> > > Please update the following components: Tabs
> > > ```
> > 
> > 
> > AntForm is having same warnings
> 
> this problem will only happen to react v16.9
> **through source code**
> 
> [ react v16.8.6](https://github.com/facebook/react/blob/16.8.6/packages/shared/ReactFeatureFlags.js)
> `export const warnAboutDeprecatedLifecycles = false;`
> 
> [ react v16.9.0](https://github.com/facebook/react/blob/master/packages/shared/ReactFeatureFlags.js)
> `export const warnAboutDeprecatedLifecycles = true;`
> 
> So the best solution now is to use react v16.8.6

excellent

**Comment by saostad** [comment]
> @Stephen2 , we are working on the 4.0 which will all use new life cycle method. For current 3.x, may more wish social volunteer to help on this.

most of the users are relying on 3.x. maybe it's better to put a little effort on 3.x to fix emergency problems like this to make sure the story of Angular 2 does not happen for this library.
The angular team was very excited to just go ahead and publish their new awesome thing, but people need reliability first!

**Comment by N00nDay** [comment]
Typography is also throwing this error

**Comment by cadrimiranda** [comment]
I want to use this library in my application, but i want this warnings fixeds because the project is to big and we cant handle this warnings in next version of react :(

**Comment by jackiehluo** [comment]
Has anyone found a way to suppress these warnings (temporarily) in their project while they wait on these fixes? They're adding so much noise to our tests right now.

**Comment by samayo** [comment]
@N00nDay Did you find a fix for this?

**Comment by ostrebler** [comment]
Please remove these warnings !

**Comment by ostrebler** [comment]
While waiting for the fixes, you can do the following to hide JUST these warnings : 

```javascript
const doWarn = window.console.warn
window.console.warn = (...args) => {
  if(typeof args[0] === 'string' && args[0].startsWith('Warning: componentWillReceiveProps has been renamed'))
    doWarn('Oh, yeah, that warning again.')
  else doWarn(...args)
}
```

**Comment by Brhernann** [comment]
> @jackiehluo and others
> While waiting for the fixes, you can do the following to hide JUST these warnings :
> 
> ```js
> const doWarn = window.console.warn
> window.console.warn = (...args) => {
>   if(typeof args[0] !== 'string' || !args[0].startsWith('Warning: componentWillReceiveProps has been renamed'))
>     doWarn(...args)
>   // Uncomment the following line if you still want a little reminder :
>   // else doWarn('Oh, yeah, that warning again.')
> }
> ```

Hi, where i can paste this code ?

**Comment by juicylevel** [comment]
Animate Animate Animate Animate Animate Animate

**Comment by uniquexiaobai** [comment]
Come on !

**Comment by dancerphil** [comment]
> may more wish social volunteer to help on this.

Tried and gave up migrating Animate. 🤣

**Comment by afc163** [comment]
In favor of react-component/form#355 and react-component/animate#82, we finally fixed all react lifecycle warnings!!! ✌️😊🎉

ref https://github.com/ant-design/ant-design/pull/19643

**Comment by tranducit95** [comment]
![Screenshot from 2019-11-11 14-37-51](https://user-images.githubusercontent.com/26267956/68569525-e14dde80-0490-11ea-8409-75f8e953b584.png)
How to fix it?

**Comment by jackiehluo** [comment]
@afc163 Does the latest release (`3.25.1`) include this fix?

**Comment by damiangreen** [comment]
No it doesnt

**Comment by blushingpenguin** [comment]
make sure you update all of the dependencies of ant too as a number of the fixes are in external components (rc-form, rc-animate, etc)

**Comment by damiangreen** [comment]
Ah yes, you're right , the final thing i had to do was change react-dom to `"react-dom": "npm:@hot-loader/react-dom",` to get rid of the final table warnings

**Comment by pavolgolias** [comment]
@afc163 - I updated completely all dependencies in project and I still got warnings for some components: Redirect, Trigger.
Did I forget something?

**Comment by IsaaX** [comment]
@pavolgolias i was receiving the same result as you. To fix it, in my yarn.lock file I deleted all the `rc-*` components since they were out of date and forced them to be reinstalled with latest versions (cautious with this approach, I have specs so everything works as before for me)

**Comment by pavolgolias** [comment]
I deleted package-lock.json and all node_modules, reinstalled the dependencies but I got still the same result. I suppose I am going to wait for new releases of rc-* components (not the beta ones) and for update of antd required components list @afc163.

**Comment by afc163** [comment]
> I suppose I am going to wait for new releases of rc-* components (not the beta ones) and for update of antd required components list

Current rc-* components have already resolved react lifecycle warning now. If you still meet this problem, could you provide a reproducible repo and steps to reproduce?

**Comment by Indigo744** [comment]
@pavolgolias  @afc163 the offending component is actually `TreeSelect`. 

I went ahead and made a repo from scratch (with `create-react-app`): https://github.com/Indigo744/antdesign-react-lifecycle-issue

Thank you

**Comment by zombieJ** [comment]
@Indigo744 , thanks for report. Updated, please check.

**Comment by blushingpenguin** [comment]
There are some duplicate components required with these fixes -- they don't seem to cause any issues, but multiple copies of rc-animate, rc-tree and rc-trigger are now required:

[antd 3.2.6](https://github.com/ant-design/ant-design/blob/master/package.json) references:
rc-animate "^2.10.2"
rc-tree-select "~2.9.1"
rc-trigger "^2.6.2"
rc-tree "~2.1.0"

[rc-tree-select@~2.9.1](https://github.com/react-component/tree-select/blob/2.9.x/package.json) references:
rc-trigger "^3.0.0-rc.2"

[rc-trigger@^3.0.0-rc.2](https://github.com/react-component/trigger/blob/997e84d99e7da344d388a2373090742a18723536/package.json) references:
rc-animate "^3.0.0-rc.1"

[rc-tree-select@~2.9.1](https://github.com/react-component/tree-select/blob/2.9.x/package.json) references:
rc-tree "~2.0.0"

**Comment by Indigo744** [comment]
Thank you @zombieJ  @afc163  I can confirm the warning is gone 😸 

However, the remarks from @blushingpenguin is interesting. If multiple version are used, does this mean a larger bundle size?

**Comment by pavolgolias** [comment]
Thanks for the update. I got still one warning left, but I found out it is another lib and not the antd.

**Comment by Sparkenstein** [comment]
Reading through the comments and status it looks like all the components have migrated, But I am still getting the same warning for InnerSlider and Slider components. attaching log and screenshot below

```sh
Warning: Using UNSAFE_componentWillMount in strict mode is not recommended and may indicate bugs in your code. See https://fb.me/react-unsafe-component-lifecycles for details.

* Move code with side effects to componentDidMount, and set initial state in the constructor.

Please update the following components: InnerSlider, Slider
```
And
```sh
Warning: Using UNSAFE_componentWillReceiveProps in strict mode is not recommended and may indicate bugs in your code. See https://fb.me/react-unsafe-component-lifecycles for details.

* Move data fetching code or side effects to componentDidUpdate.
* If you're updating state whenever props change, refactor your code to use memoization techniques or move it to static getDerivedStateFromProps. Learn more at: https://fb.me/react-derived-state

Please update the following components: InnerSlider
```
I am using `React 16.13.1` and `AntD 4.3.3`

**Comment by boriscy** [comment]
> Reading through the comments and status it looks like all the components have migrated, But I am still getting the same warning for InnerSlider and Slider components. attaching log and screenshot below
> 
> ```shell
> Warning: Using UNSAFE_componentWillMount in strict mode is not recommended and may indicate bugs in your code. See https://fb.me/react-unsafe-component-lifecycles for details.
> 
> * Move code with side effects to componentDidMount, and set initial state in the constructor.
> 
> Please update the following components: InnerSlider, Slider
> ```
> 
> And
> 
> ```shell
> Warning: Using UNSAFE_componentWillReceiveProps in strict mode is not recommended and may indicate bugs in your code. See https://fb.me/react-unsafe-component-lifecycles for details.
> 
> * Move data fetching code or side effects to componentDidUpdate.
> * If you're updating state whenever props change, refactor your code to use memoization techniques or move it to static getDerivedStateFromProps. Learn more at: https://fb.me/react-derived-state
> 
> Please update the following components: InnerSlider
> ```
> 
> I am using `React 16.13.1` and `AntD 4.3.3`

Same problem with Modal and antd 4.3.4

**Comment by shmaram** [comment]
Same problem still with React 16.13.1 and AndD 4.5.4
is there any plan to fix this? 

![image](https://user-images.githubusercontent.com/12017112/90765761-28f07900-e2f3-11ea-8440-2740ef650288.png)

**Comment by xeontem** [comment]
same for me. Should we reopen the ticket?
